clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/AbstractPubSubBasedMessageSender.java

```diff
@@ -0,0 +1,306 @@
+     *
+     * @param operationName The operation name to set for the span.
+     * @param topic The topic to which the message is sent.
+     * @param tenantId The tenant identifier related to the operation.
+     * @param deviceId The device identifier related to the operation.
+     * @param referenceType The type of reference towards the given span context.
+     * @param context The span context to set as parent and to derive the sampling priority from (may be null).
+     * @return The new span.
+     * @throws NullPointerException if tracer or topic is {@code null}.
+     */
+    protected Span startSpan(final String operationName, final String topic, final String tenantId,
+            final String deviceId, final String referenceType, final SpanContext context) {
+        Objects.requireNonNull(operationName);
+        Objects.requireNonNull(topic);
+        Objects.requireNonNull(referenceType);
+        return newSpan(tracer, operationName, topic, referenceType, context)
+                .setTag(TracingHelper.TAG_TENANT_ID.getKey(), tenantId)
+                .setTag(TracingHelper.TAG_DEVICE_ID.getKey(), deviceId);
+    }
+
+    private PubSubPublisherClient getOrCreatePublisher(final String topic, final String tenantId) {
+        return publisherFactory.getOrCreatePublisher(topic, projectId, tenantId);
+    }
+
+    private void logPubSubMessageId(final Span span, final String topic, final String messageId) {
+        log.trace("message published to PubSub [topic: {}, id: {}]", topic, messageId);
+        span.log("message published to PubSub");
+
+        Tags.HTTP_STATUS.set(span, HttpURLConnection.HTTP_ACCEPTED);
+    }
+
+    private void logPubSubMessage(final Span span, final PubsubMessage message, final String topic,
+            final String tenantId) {
+        final String attributesAsString = message.getAttributesMap()
+                .keySet()
+                .stream()
+                .map(key -> key + "=" + message.getAttributesMap().get(key))
+                .collect(Collectors.joining(",", "{", "}"));
+        log.trace("producing message [topic: {}, tenant: {}, key: {}, timestamp: {}, attributes: {}]",
+                topic, tenantId, message.getOrderingKey(), message.getPublishTime(), attributesAsString);
+
+        span.log("publishing message with headers: " + attributesAsString);
+    }
+
+    private void logError(
+            final Span span,
+            final String topic,
+            final String tenantId,
+            final String deviceId,
+            final Throwable cause) {
+        log.debug("sending message failed [topic: {}, key: {}, tenantId: {}, deviceId: {}]", topic, deviceId, tenantId,
+                deviceId, cause);
+
+        Tags.HTTP_STATUS.set(span, getErrorCode());
+        TracingHelper.logError(span, cause);
+    }
+
+    private int getErrorCode() {
+        return HttpURLConnection.HTTP_UNAVAILABLE;
+    }
```

Comment:
this does not seem to serve any real purpose except for making the code where the exception is being thrown harder to read/understand. FMPOV we should simply remove this method ...

Reply:
Agree, removed this method